### PR TITLE
Fix nested slicing arguments and nested variables

### PR DESCRIPTION
**Pull request by yahelnachum**

### CODE REVIEW

File: src/generate-query.ts [hunk 1]
```diff
@@ -610,6 +669,145 @@ function getArgsAndVars(
   }
 }
 
+function getListSizeDirective(field: FieldDefinitionNode): DirectiveNode {
+  return field?.directives.find((dir) => dir.name.value === 'listSize')
+}
+
+function getSlicingArguments(
```
Comment (Alan-Cha) [inline src/generate-query.ts:676]: I think this function is a bit strange. It is doing two things, 1) getting the list size directive (if it isn't provided and the `field` is) and 2) extracting the slicing arguments. I think the two functions should be split apart. We should have a `getListSizeDirective()` function (which we already have) and a `getSlicingArgumentsFromListSIzeDirective()` function that just takes a list size directive. Then in the code, if we already have the list size directive, we just call the latter, and if we only have a field, then we call both.

File: src/generate-query.ts [hunk 2]
```diff
@@ -610,6 +669,145 @@ function getArgsAndVars(
...
+    (arg) => arg.name.value === 'slicingArguments'
+  )
+
+  if (slicingArgumentsArg && slicingArgumentsArg.value.kind !== 'ListValue')
+    return { slicingArgumentsArg: undefined, listSizeDirective: undefined }
+
+  return { slicingArgumentsArg, listSizeDirective }
+}
+
+function getRequireOneSlicingArgument(
```
Comment (Alan-Cha) [inline src/generate-query.ts:699]: I have similar thoughts with this function. I think it should just be`getRequireOneSlicingArgumentFromListSizeDirective()` or better yet, is there a way to combine both this and `getSlicingArguments()` so it just extracts all the relevant information from a list size directive?

File: src/generate-query.ts (hunk 1 above)
Comment (yahelnachum) [reply src/generate-query.ts:676]: Yes, that makes sense. I will change the code as you mentioned.

File: src/generate-query.ts (hunk 2 above)
Comment (yahelnachum) [reply src/generate-query.ts:699]: We could combine them, but I might like the piecemeal approach more. I'll switch it to a simpler function then.

File: src/generate-query.ts [hunk 3]
```diff
@@ -509,9 +506,52 @@ function getMissingSlicingArg(
...
+  schema: GraphQLSchema
+) {
+  if (type?.astNode?.kind === Kind.INPUT_OBJECT_TYPE_DEFINITION) {
+    return type.astNode.fields.find((field) => {
+      let ret = getMissingSlicingArgHelper(
+        tokenizedPath.slice(1),
+        field,
+        schema
+      )
+      if (ret) return ret
```
Comment (Alan-Cha) [inline src/generate-query.ts:533]: You can just return the result. I don't think you need to check if it exists.

File: src/generate-query.ts [hunk 4]
```diff
@@ -509,9 +506,52 @@ function getMissingSlicingArg(
...
+      if (ret) return arg
+    })
+  })
+}
+
+function getMissingSlicingArgHelper1(
+  tokenizedPath: string[],
+  type: GraphQLNamedType,
+  schema: GraphQLSchema
+) {
```
Comment (Alan-Cha) [inline src/generate-query.ts:525]: Can we add a return type?

File: src/generate-query.ts [hunk 5]
```diff
@@ -509,9 +506,52 @@ function getMissingSlicingArg(
...
+        slicingArgument.split('.'),
+        arg,
+        schema
+      )
+      if (ret) return arg
+    })
+  })
+}
+
+function getMissingSlicingArgHelper1(
```
Comment (Alan-Cha) [inline src/generate-query.ts:521]: You know me, I'm not a fan of this name 😅

File: src/generate-query.ts [hunk 6]
```diff
@@ -509,9 +506,52 @@ function getMissingSlicingArg(
   if (usesSlicingArg) return null
 
   // Return the first slicing arguments:
```
Comment (Alan-Cha) [inline src/generate-query.ts:508]: I'm having a bit of trouble understanding what the following is doing. Could you remind me? And while we're at it, we can improve this comment so it's more helpful next time.

File: src/generate-query.ts [hunk 7]
```diff
@@ -509,9 +506,52 @@ function getMissingSlicingArg(
...
+  else if (field.name.value !== tokenizedPath[0]) return undefined
+  else if (tokenizedPath.length === 1) return field
+
+  if (
+    field.kind !== Kind.INPUT_VALUE_DEFINITION ||
+    field.type.kind !== Kind.NAMED_TYPE
   )
+    return undefined
+
+  let type = schema.getType(field.type.name.value)
```
Comment (Alan-Cha) [inline src/generate-query.ts:553]: `const`

File: src/generate-query.ts [hunk 8]
```diff
@@ -509,9 +506,52 @@ function getMissingSlicingArg(
...
-  return field.arguments.find((arg) =>
-    slicingArguments.includes(arg.name.value)
+  return field.arguments.find((arg) => {
+    return slicingArguments.find((slicingArgument) => {
+      let ret = getMissingSlicingArgHelper(
+        slicingArgument.split('.'),
+        arg,
+        schema
+      )
+      if (ret) return arg
```
Comment (Alan-Cha) [inline src/generate-query.ts:516]: This is strange to me. The callback function inside a `find()` iterator should return a boolean but you are return `arg`?

File: src/generate-query.ts (hunk 3 above)
Comment (yahelnachum) [reply src/generate-query.ts:533]: sounds good

File: src/generate-query.ts (hunk 7 above)
Comment (yahelnachum) [reply src/generate-query.ts:553]: done

File: src/generate-query.ts (hunk 8 above)
Comment (yahelnachum) [reply src/generate-query.ts:516]: fixed

File: src/generate-query.ts (hunk 4 above)
Comment (yahelnachum) [reply src/generate-query.ts:525]: fixed

File: src/generate-query.ts (hunk 6 above)
Comment (yahelnachum) [reply src/generate-query.ts:508]: added better comment

File: src/generate-query.ts (hunk 5 above)
Comment (yahelnachum) [reply src/generate-query.ts:521]: fixed

File: src/generate-query.ts (hunk 6 above)
Comment (yahelnachum) [reply src/generate-query.ts]:
```
  // Returns the first slicing argument
  // if slicing argument can be split on '.'
  // then we have a nested slicing argument,
  // and we need to follow the path all the way
  // down to the leaf input field definition
```

File: src/generate-query.ts [hunk 9]
```diff
@@ -508,10 +505,56 @@ function getMissingSlicingArg(
...
+  field: InputValueDefinitionNode,
+  schema: GraphQLSchema
+): InputValueDefinitionNode {
+  if (tokenizedPath.length < 1) return undefined
+  else if (field.name.value !== tokenizedPath[0]) return undefined
+  else if (tokenizedPath.length === 1) return field
+
+  if (
+    field.kind !== Kind.INPUT_VALUE_DEFINITION ||
+    field.type.kind !== Kind.NAMED_TYPE
```
Comment (Alan-Cha) [inline src/generate-query.ts:552]: If the type isn't a named type, then should we try to unwrap it? Or is this not a case?

File: src/generate-query.ts [hunk 10]
```diff
@@ -508,10 +505,56 @@ function getMissingSlicingArg(
...
-  // Return the first slicing arguments:
-  return field.arguments.find((arg) =>
-    slicingArguments.includes(arg.name.value)
+  // Returns the first slicing argument
+  // if slicing argument can be split on '.'
+  // then we have a nested slicing argument,
+  // and we need to follow the path all the way
+  // down to the leaf input field definition
+  return field.arguments.find((arg) => {
+    return slicingArguments.find((slicingArgument) => {
```
Comment (Alan-Cha) [inline src/generate-query.ts:514]: Can you remind me why we need to follow all the way down to the leaf input field definition when we are just returning the top level argument? I feel like there's something strange about this whole thing or I'm just not understanding it correctly.

Comment (yahelnachum) [reply src/generate-query.ts:514]: We need to follow to make sure that path down actually exists. I guess, what you are saying is that this repo is not about actually verifying that the path exists, but about returning the top level argument that matches?

Comment (Alan-Cha) [reply src/generate-query.ts:514]: Yeah, is this library responsible for checking if the slicing argument is valid or if the list size directive is valid as a whole? I'm not sure.

Comment (yahelnachum) [reply src/generate-query.ts:512]: I agree, it is not this repo's responsibility. I will remove this, make it a single level match.

File: src/generate-query.ts [hunk 11]
```diff
@@ -610,6 +670,121 @@ function getArgsAndVars(
...
+    (requireOneSlicingArg.value as BooleanValueNode).value === true
+  ) {
+    slicingArgumentsPaths = slicingArgumentsPaths.slice(0, 1)
+  }
+  const slicingArgumentsTokenizedPaths = slicingArgumentsPaths.map((value) =>
+    value.split('.')
+  )
+
+  return slicingArgumentsTokenizedPaths.reduce((res, tokenizedPath) => {
+    return merge.recursive(
```
Comment (Alan-Cha) [inline src/generate-query.ts:693]: Doesn't this assume that all slicing arguments are input object types? That's not necessarily true right? It could just be a simple integer right?

File: src/generate-query.ts (hunk 9 above)
Comment (yahelnachum) [reply src/generate-query.ts]: removed function so this is outdated, resolving convo

File: src/generate-query.ts [hunk 12]
```diff
@@ -610,6 +670,121 @@ function getArgsAndVars(
...
+    (value) => {
+      if (value.kind === 'StringValue') return value.value
+    }
+  )
+  const requireOneSlicingArg = getRequireOneSlicingArgument(listSizeDirective)
+  if (
+    !requireOneSlicingArg ||
+    (requireOneSlicingArg.value as BooleanValueNode).value === true
+  ) {
+    slicingArgumentsPaths = slicingArgumentsPaths.slice(0, 1)
```
Comment (Alan-Cha) [inline src/generate-query.ts:686]: Why are we slicing here?

Comment (yahelnachum) [reply src/generate-query.ts:686]: since require only one slicing is defaulted to true (or if it is present and set to true) then we only can add a single slicing arg in the query / variables. If we have 2 slicing args and that is set to true then its an invalid gql request

File: src/generate-query.ts (hunk 10 above)
Comment (yahelnachum) [reply src/generate-query.ts:512]: resolving convo, let me know if you disagree with how I went about it and have more suggestions

File: src/generate-query.ts (hunk 11 above)
Comment (yahelnachum) [reply src/generate-query.ts:693]: ahh yea, you are right. Good catch, I will fix this soon.

Comment (yahelnachum) [reply src/generate-query.ts:707]: fixed, resolving conversation

File: src/generate-query.ts (hunk 12 above)
Comment (yahelnachum) [reply src/generate-query.ts:689]: resolving convo, if you have more questions you can reopen or recomment as needed, thanks

File: src/generate-query.ts [hunk 13]
```diff
@@ -610,6 +638,170 @@ function getArgsAndVars(
...
+  }
+  const slicingArgumentsTokenizedPaths = slicingArgumentsPaths.map((value) =>
+    value.split('.')
+  )
+
+  if (
+    slicingArgumentsPaths.length === 1 &&
```
Comment (Alan-Cha) [inline src/generate-query.ts:720]: Why is `slicingArgumentsPaths.length === 1` necessary?

Comment (yahelnachum) [reply src/generate-query.ts:720]: If requireOneSlicingArg is false and
```
const slicingArgumentsArg = getSlicingArguments(listSizeDirective)
```
returns no slicing args then the paths could be 0 so we might not be able to index into the first one

Comment (yahelnachum) [reply src/generate-query.ts:720]: also if only 1 is in the list and it has only 1 token then we will be returning a simple var value back

Comment (Alan-Cha) [reply src/generate-query.ts:720]: I was wondering why it isn't `slicingArgumentsPaths.length > 0`. Doesn't this behave strangely if there is more than one path?

File: src/generate-query.ts [hunk 14]
```diff
@@ -586,9 +593,30 @@ function getArgsAndVars(
...
+        // Enum values can be randomly selected
         variableValues[varName] = getRandomEnum(argType)
       } else if (config.providePlaceholders) {
-        variableValues[varName] = getDefaultArgValue(arg.type)
+        variableValues[varName] = getDefaultArgValue(schema, config, arg.type)
+
+        // Add default values for unfulfilled listSize directive if applicable
+        const listSizeDirective = getListSizeDirective(field)
+        if (listSizeDirective) {
+          const value = getListSizeDirectiveDefaultValue(
```
Comment (Alan-Cha) [inline src/generate-query.ts:604]: I think this doesn't work as intended with `requireOneSlicingArg`. With the current way that `getListSizeDirectiveDefaultValue` is being called, it doesn't matter whether `requireOneSlicingArg` is set to true or not, we can only add at most one slicing argument. What do you think?

Comment (Alan-Cha) [reply src/generate-query.ts:604]: Let me rephrase, it works with multiple input object types but if there are multiple integer slicing args, this doesn't work right

File: src/generate-query.ts (hunk 13 above)
Comment (yahelnachum) [reply src/generate-query.ts:720]: hmmm... yea I guess there is a disconnect between the slicingargs list (could contain more than one argument root) and the caller which just wants a value for one argument

File: src/generate-query.ts (hunk 14 above)
Comment (yahelnachum) [reply src/generate-query.ts:604]: yea I agree, I think this is a similar disconnect to the comment I made here
>hmmm... yea I guess there is a disconnect between the slicingargs list (could contain more than one argument root) and the caller which just wants a value for one argument
https://github.com/IBM/graphql-query-generator/pull/38/files#r650224767

File: src/generate-query.ts (hunk 13 above)
Comment (yahelnachum) [reply src/generate-query.ts:720]: @Alan-Cha, I pushed some new tests that I thought would fail with the code we were discussing (multiple single level slicing arguments, requireOneSlicingArg: false). Surprisingly they passed. 

The multiple single level slicing arguments one works because requireOneSlicingArg is defaulted to true so we only consider a single slicing arg when creating variables.

The requireOneSlicingArg: false works because an earlier call to getMissingSlicingArgs does not return them since setting it to false means they are never required.

So I think there is some dead code in there, but it all works based on what I have thought of. Let me know if you think there is a test that will break it.

File: src/generate-query.ts [hunk 15]
```diff
@@ -610,6 +638,170 @@ function getArgsAndVars(
...
+      res,
+      getListSizeDirectiveDefaultValueHelper(
+        tokenizedPath,
+        typeNode,
+        config,
+        schema
+      )
+    )
+  }, {})
+}
```
Comment (yahelnachum) [inline src/generate-query.ts:736]: I think this is dead code, since we either `requireOneSlicingArgument: true` so we only use the return right before or `requireOneSlicingArgument: false` so `getMissingSlicingArguments` never returns anything as "required".

File: src/generate-query.ts (hunk 13 above)
Comment (yahelnachum) [reply src/generate-query.ts]: @Alan-Cha I addressed this specific comment in the latest push. I will mark this conversation as resolved.

File: src/generate-query.ts [hunk 16]
```diff
@@ -707,32 +709,24 @@ function getListSizeDirectiveDefaultValue(
 
   const requireOneSlicingArg = getRequireOneSlicingArgument(listSizeDirective)
   if (
-    !requireOneSlicingArg ||
-    (requireOneSlicingArg.value as BooleanValueNode).value === true
+    requireOneSlicingArg &&
+    (requireOneSlicingArg.value as BooleanValueNode).value === false
   ) {
-    slicingArgumentsPaths = slicingArgumentsPaths.slice(0, 1)
+    throw new Error(
```
Comment (Alan-Cha) [inline src/generate-query.ts:715]: I don't think it is a good idea to throw an error when this is set to false.

File: src/generate-query.ts [hunk 17]
```diff
@@ -608,13 +608,15 @@ function getArgsAndVars(
...
-          if (typeof value !== 'object') {
-            variableValues[varName] = value
-          } else {
-            variableValues[varName] = merge.recursive(
-              value,
-              variableValues[varName]
-            )
+          if (value !== undefined) {
+            if (typeof value !== 'object') {
+              variableValues[varName] = value
```
Comment (Alan-Cha) [inline src/generate-query.ts:613]: What I expected you to do was, instead of adding the `@listSize` directive default values here (or more properly, single default value), pass `variableValues` into `getListSizeDirectiveDefaultValue()`. That way, you can add multiple values into the `variableValues` as you iterate through all of slicing arguments if `requireOneSlicingArg` is set to `false`.

File: src/generate-query.ts [hunk 18]
```diff
@@ -707,32 +709,24 @@ function getListSizeDirectiveDefaultValue(
...
+  const slicingArgumentsTokenizedPath = slicingArgumentsPaths[0].split('.')
 
-  if (
-    slicingArgumentsPaths.length === 1 &&
-    slicingArgumentsTokenizedPaths[0].length === 1
-  )
+  if (slicingArgumentsTokenizedPath.length === 1)
     return getDefaultArgValue(schema, config, typeNode)
 
-  return slicingArgumentsTokenizedPaths.reduce((res, tokenizedPath) => {
```
Comment (Alan-Cha) [inline src/generate-query.ts]: I think what you had here is pretty sensible (i.e. slicing `slicingArgumentsTokenizedPaths` when `requireOneSlicingArg` is set to true, otherwise iterating over all the paths using a `reduce`).

What you could do to add multiple default values (if `requireOneSlicingArg` is set to true`) is, 1) pass in the `variableValues` and 2) instead of doing a `reduce()`, you can just do a `forEach()` and in the call back, add the value into `variableValues`.

Comment (Alan-Cha) [reply src/generate-query.ts]: Instead of immediately adding new entries to the `variableValues`, I think we should consider checking if 1) the key already exists in which case we should throw an error (because our code is supposed to ignore variables that have already been provided a value) and 2) attempt a merge when the type of the variable is an input object

File: src/generate-query.ts (hunk 16 above)
Comment (yahelnachum) [reply src/generate-query.ts:715]: should it be an `assert` instead then? Since it would be an internal programming error to get to this function otherwise?

Comment (Alan-Cha) [reply src/generate-query.ts:715]: I don't think we should break when we are using `@listSize` directive in a very ordinary way, but what I'm really asking is if it is possible to address this problem with what I discussed below.
